### The following processes will be conducted in this data wrangling section:

### Reading the raw data from a CSV file

### Display data samples to understand its structure

### Data Cleaning: Identify and handle missing values

### Feature Engineering: Create new features based on existing ones (e.g., calculate mean income per zip code)

### Data Filtering: Filter out properties that are not relevant to the analysis (e.g., outliers or properties outside Austin city limits)

### # Feature Engineering: Create new features based on existing ones (e.g., calculate price per square foot)

### Data Aggregation: Calculate average price per zip code

### Data Merging: Combining housing data with demographic data

### Data Visualization: Visualize the data to gain insights and identify trends

### After data wrangling, the cleaned and transformed data can be used for further analysis and modeling.


In [1]:
# First, import the relevant modules and packages
import pandas as pd
import numpy as np
#from matplotlib import pyplot
import matplotlib.pyplot as plt

In [2]:
# Import Austin_TX_House_Listings_data_v1.csv & us_income_zipcode.csv
df1 = pd.read_csv('../01_raw_data/AustinTXHouseListingsDataV1.csv', low_memory=False)

In [3]:
# Displaying the first few rows of the raw data to understand its structure
df1.head()

Unnamed: 0.1  Unnamed: 0 address_city          address_neighborhood  \
0             0          15       Austin                           NaN   
1             1          16       Austin                           NaN   
2             2          17       Austin                           NaN   
3             3          18       Austin                           NaN   
4             4          19       Austin  Steiner Ranch-Lakewood Hills   

  address_state    address_streetAddress address_subdivision  address_zipcode  \
0            TX     12801 Wooded Lake Ct                 NaN            78732   
1            TX   904 Lakewood Hills Ter                 NaN            78732   
2            TX        13701 Montview Dr                 NaN            78732   
3            TX   700 Lakewood Hills Ter                 NaN            78732   
4            TX  1008 Lakewood Hills Ter                 NaN            78732   

   bathrooms  bedrooms  ...  latest_salemonth latest_saleyear  \
0        5.0       4.0  ...                 2            2018   
1        5.0       5.0  ...                 8            2020   
2        3.5       NaN  ...                 3            2018   
3        4.0       4.0  ...                 6            2018   
4        4.0       4.0  ...                 8            2019   

  latest_price_source  numOfAccessibilityFeatures  numOfAppliances  numPhotos  \
0      Agent Provided                           0                3         39   
1     Broker Provided                           0                3         58   
2      Agent Provided                           0                0          1   
3      Agent Provided                           0                1         40   
4     Broker Provided                           0                4         40   

   numPriceChanges                                              photo  \
0                1  https://photos.zillowstatic.com/fp/911f9a59acb...   
1                2  https://photos.zillowstatic.com/fp/c6346c4a39d...   
2                1  https://maps.googleapis.com/maps/api/streetvie...   
3                4  https://photos.zillowstatic.com/fp/805b5a6b748...   
4                7  https://photos.zillowstatic.com/fp/a09cfad45a9...   

                                           imagepath  \
0  /content/gdrive/MyDrive/zillow-images/70352485...   
1  /content/gdrive/MyDrive/zillow-images/70352465...   
2  /content/gdrive/MyDrive/zillow-images/83823478...   
3  /content/gdrive/MyDrive/zillow-images/70352478...   
4  /content/gdrive/MyDrive/zillow-images/70352461...   

                                           cleanpath  
0  70352485_911f9a59acb6fe5fd909538a685e2a0c-p_f.jpg  
1  70352465_c6346c4a39d1f87578ee1529329d9296-p_f.jpg  
2  83823478_streetviewlocation13701MontviewDr2CAu...  
3  70352478_805b5a6b748b1d9cffeb93a4c41fedb0-p_f.jpg  
4  70352461_a09cfad45a949f7195c9540224f51e49-p_f.jpg  

[5 rows x 736 columns]

In [4]:
df1.shape

(16482, 736)

It looks like our data has more features (columns) than we need

In [5]:
#Analyze the summary of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16482 entries, 0 to 16481
Columns: 736 entries, Unnamed: 0.1 to cleanpath
dtypes: bool(13), float64(227), int64(10), object(486)
memory usage: 91.1+ MB


In [6]:
df1.select_dtypes(include="object").count()

address_city             16482
address_neighborhood       431
address_state            16482
address_streetAddress    16482
address_subdivision        175
                         ...  
latest_saledate          16482
latest_price_source      16482
photo                    16482
imagepath                16482
cleanpath                16482
Length: 486, dtype: int64

In [7]:
list(df1.select_dtypes(include="object"))

['address_city',
 'address_neighborhood',
 'address_state',
 'address_streetAddress',
 'address_subdivision',
 'description',
 'homeStatus',
 'resoFactsStats_accessibilityFeatures_0',
 'resoFactsStats_accessibilityFeatures_1',
 'resoFactsStats_accessibilityFeatures_2',
 'resoFactsStats_accessibilityFeatures_3',
 'resoFactsStats_accessibilityFeatures_4',
 'resoFactsStats_accessibilityFeatures_5',
 'resoFactsStats_accessibilityFeatures_6',
 'resoFactsStats_accessibilityFeatures_7',
 'resoFactsStats_additionalParcelsDescription',
 'resoFactsStats_appliances_0',
 'resoFactsStats_appliances_1',
 'resoFactsStats_appliances_2',
 'resoFactsStats_appliances_3',
 'resoFactsStats_appliances_4',
 'resoFactsStats_appliances_5',
 'resoFactsStats_appliances_6',
 'resoFactsStats_appliances_7',
 'resoFactsStats_appliances_8',
 'resoFactsStats_appliances_9',
 'resoFactsStats_appliances_10',
 'resoFactsStats_appliances_11',
 'resoFactsStats_appliances_12',
 'resoFactsStats_architecturalStyle',
 'resoFact

In [8]:
list(df1['lotSize'])

['0.36 Acres',
 '0.66 Acres',
 '0.50 Acres',
 '0.49 Acres',
 '0.48 Acres',
 '7,797 sqft',
 '0.41 Acres',
 '0.47 Acres',
 '2.67 Acres',
 '0.28 Acres',
 '0.30 Acres',
 '0.30 Acres',
 '0.30 Acres',
 '0.59 Acres',
 '8,232 sqft',
 '8,276 sqft',
 '0.32 Acres',
 '7,405 sqft',
 '0.52 Acres',
 '8,842 sqft',
 '0.29 Acres',
 '0.31 Acres',
 '8,363 sqft',
 '0.28 Acres',
 '0.27 Acres',
 '7,971 sqft',
 '9,931 sqft',
 '10,497 sqft',
 '2.68 Acres',
 '7,492 sqft',
 '0.37 Acres',
 '0.36 Acres',
 '0.36 Acres',
 '7,753 sqft',
 '0.37 Acres',
 '0.44 Acres',
 '4.18 Acres',
 '1.34 Acres',
 '5,662 sqft',
 '0.29 Acres',
 '10,018 sqft',
 '2.35 Acres',
 '0.26 Acres',
 '8,929 sqft',
 '0.34 Acres',
 '1.07 Acres',
 '0.28 Acres',
 '0.59 Acres',
 '1.36 Acres',
 '3.96 Acres',
 '0.34 Acres',
 '9,888 sqft',
 '0.32 Acres',
 '1.86 Acres',
 '6,316 sqft',
 '0.87 Acres',
 '9,670 sqft',
 '0.76 Acres',
 '0.50 Acres',
 '1.31 Acres',
 '10,367 sqft',
 '0.48 Acres',
 '1.25 Acres',
 '0.40 Acres',
 '3.54 Acres',
 '6,011 sqft',
 '1.52 

In [9]:
df1[['lotSize_value', 'lotSize_unit']] = df1['lotSize'].str.split(' ', expand=True)
df1.head(3)

Unnamed: 0.1  Unnamed: 0 address_city address_neighborhood address_state  \
0             0          15       Austin                  NaN            TX   
1             1          16       Austin                  NaN            TX   
2             2          17       Austin                  NaN            TX   

    address_streetAddress address_subdivision  address_zipcode  bathrooms  \
0    12801 Wooded Lake Ct                 NaN            78732        5.0   
1  904 Lakewood Hills Ter                 NaN            78732        5.0   
2       13701 Montview Dr                 NaN            78732        3.5   

   bedrooms  ...  latest_price_source numOfAccessibilityFeatures  \
0       4.0  ...       Agent Provided                          0   
1       5.0  ...      Broker Provided                          0   
2       NaN  ...       Agent Provided                          0   

  numOfAppliances  numPhotos  numPriceChanges  \
0               3         39                1   
1               3         58                2   
2               0          1                1   

                                               photo  \
0  https://photos.zillowstatic.com/fp/911f9a59acb...   
1  https://photos.zillowstatic.com/fp/c6346c4a39d...   
2  https://maps.googleapis.com/maps/api/streetvie...   

                                           imagepath  \
0  /content/gdrive/MyDrive/zillow-images/70352485...   
1  /content/gdrive/MyDrive/zillow-images/70352465...   
2  /content/gdrive/MyDrive/zillow-images/83823478...   

                                           cleanpath lotSize_value  \
0  70352485_911f9a59acb6fe5fd909538a685e2a0c-p_f.jpg          0.36   
1  70352465_c6346c4a39d1f87578ee1529329d9296-p_f.jpg          0.66   
2  83823478_streetviewlocation13701MontviewDr2CAu...          0.50   

  lotSize_unit  
0        Acres  
1        Acres  
2        Acres  

[3 rows x 738 columns]

In [10]:
df1.columns

Index(['Unnamed: 0.1', 'Unnamed: 0', 'address_city', 'address_neighborhood',
       'address_state', 'address_streetAddress', 'address_subdivision',
       'address_zipcode', 'bathrooms', 'bedrooms',
       ...
       'latest_price_source', 'numOfAccessibilityFeatures', 'numOfAppliances',
       'numPhotos', 'numPriceChanges', 'photo', 'imagepath', 'cleanpath',
       'lotSize_value', 'lotSize_unit'],
      dtype='object', length=738)

In [11]:
list(df1.select_dtypes(include="object"))

['address_city',
 'address_neighborhood',
 'address_state',
 'address_streetAddress',
 'address_subdivision',
 'description',
 'homeStatus',
 'resoFactsStats_accessibilityFeatures_0',
 'resoFactsStats_accessibilityFeatures_1',
 'resoFactsStats_accessibilityFeatures_2',
 'resoFactsStats_accessibilityFeatures_3',
 'resoFactsStats_accessibilityFeatures_4',
 'resoFactsStats_accessibilityFeatures_5',
 'resoFactsStats_accessibilityFeatures_6',
 'resoFactsStats_accessibilityFeatures_7',
 'resoFactsStats_additionalParcelsDescription',
 'resoFactsStats_appliances_0',
 'resoFactsStats_appliances_1',
 'resoFactsStats_appliances_2',
 'resoFactsStats_appliances_3',
 'resoFactsStats_appliances_4',
 'resoFactsStats_appliances_5',
 'resoFactsStats_appliances_6',
 'resoFactsStats_appliances_7',
 'resoFactsStats_appliances_8',
 'resoFactsStats_appliances_9',
 'resoFactsStats_appliances_10',
 'resoFactsStats_appliances_11',
 'resoFactsStats_appliances_12',
 'resoFactsStats_architecturalStyle',
 'resoFact

In [12]:
df1['lotSize_unit'].unique()

array(['Acres', 'sqft', nan], dtype=object)

In [13]:
len(df1['lotSize_value'])

16482

In [14]:
df1['lotSize_value'].info

<bound method Series.info of 0          0.36
1          0.66
2          0.50
3          0.49
4          0.48
          ...  
16477      0.54
16478     9,321
16479    10,280
16480     8,450
16481     7,797
Name: lotSize_value, Length: 16482, dtype: object>

In [15]:
df1.replace('nan', float('nan'), inplace=True)

In [16]:
df1['lotSize_unit'].unique()

array(['Acres', 'sqft', nan], dtype=object)

In [17]:
df1['lotSize_value'].replace(',', '', inplace=True)

In [18]:
df1['lotSize_value'].info

<bound method Series.info of 0          0.36
1          0.66
2          0.50
3          0.49
4          0.48
          ...  
16477      0.54
16478     9,321
16479    10,280
16480     8,450
16481     7,797
Name: lotSize_value, Length: 16482, dtype: object>

In [19]:
df1['lotSize_value'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 16482 entries, 0 to 16481
Series name: lotSize_value
Non-Null Count  Dtype 
--------------  ----- 
16154 non-null  object
dtypes: object(1)
memory usage: 128.9+ KB


In [20]:
df1['lotSize_value'] = df1['lotSize_value'].str.replace(',', '')

In [21]:
df1['lotSize_value'].info

<bound method Series.info of 0         0.36
1         0.66
2         0.50
3         0.49
4         0.48
         ...  
16477     0.54
16478     9321
16479    10280
16480     8450
16481     7797
Name: lotSize_value, Length: 16482, dtype: object>

In [22]:
df1['lotSize_value'] = pd.to_numeric(df1['lotSize_value'])

In [23]:
df1[['lotSize_unit']]

lotSize_unit
0            Acres
1            Acres
2            Acres
3            Acres
4            Acres
...            ...
16477        Acres
16478         sqft
16479         sqft
16480         sqft
16481         sqft

[16482 rows x 1 columns]

In [24]:
# Convert acres to square feet where column_unit is 'Acres'
df1['lotSize_sqft'] = np.where(df1['lotSize_unit'] == 'Acres',
                               df1['lotSize_value'] * 43560,
                               df1['lotSize_value'])
df1.head()

Unnamed: 0.1  Unnamed: 0 address_city          address_neighborhood  \
0             0          15       Austin                           NaN   
1             1          16       Austin                           NaN   
2             2          17       Austin                           NaN   
3             3          18       Austin                           NaN   
4             4          19       Austin  Steiner Ranch-Lakewood Hills   

  address_state    address_streetAddress address_subdivision  address_zipcode  \
0            TX     12801 Wooded Lake Ct                 NaN            78732   
1            TX   904 Lakewood Hills Ter                 NaN            78732   
2            TX        13701 Montview Dr                 NaN            78732   
3            TX   700 Lakewood Hills Ter                 NaN            78732   
4            TX  1008 Lakewood Hills Ter                 NaN            78732   

   bathrooms  bedrooms  ...  numOfAccessibilityFeatures numOfAppliances  \
0        5.0       4.0  ...                           0               3   
1        5.0       5.0  ...                           0               3   
2        3.5       NaN  ...                           0               0   
3        4.0       4.0  ...                           0               1   
4        4.0       4.0  ...                           0               4   

  numPhotos  numPriceChanges  \
0        39                1   
1        58                2   
2         1                1   
3        40                4   
4        40                7   

                                               photo  \
0  https://photos.zillowstatic.com/fp/911f9a59acb...   
1  https://photos.zillowstatic.com/fp/c6346c4a39d...   
2  https://maps.googleapis.com/maps/api/streetvie...   
3  https://photos.zillowstatic.com/fp/805b5a6b748...   
4  https://photos.zillowstatic.com/fp/a09cfad45a9...   

                                           imagepath  \
0  /content/gdrive/MyDrive/zillow-images/70352485...   
1  /content/gdrive/MyDrive/zillow-images/70352465...   
2  /content/gdrive/MyDrive/zillow-images/83823478...   
3  /content/gdrive/MyDrive/zillow-images/70352478...   
4  /content/gdrive/MyDrive/zillow-images/70352461...   

                                           cleanpath lotSize_value  \
0  70352485_911f9a59acb6fe5fd909538a685e2a0c-p_f.jpg          0.36   
1  70352465_c6346c4a39d1f87578ee1529329d9296-p_f.jpg          0.66   
2  83823478_streetviewlocation13701MontviewDr2CAu...          0.50   
3  70352478_805b5a6b748b1d9cffeb93a4c41fedb0-p_f.jpg          0.49   
4  70352461_a09cfad45a949f7195c9540224f51e49-p_f.jpg          0.48   

  lotSize_unit lotSize_sqft  
0        Acres      15681.6  
1        Acres      28749.6  
2        Acres      21780.0  
3        Acres      21344.4  
4        Acres      20908.8  

[5 rows x 739 columns]

In [25]:
# Analyze how the dependent variable, 'latest_price', is correlated with the rest of independent variables
dict(df1.corr()['latest_price'].abs().sort_values(ascending = False))

C:\Users\famma\AppData\Local\Temp\ipykernel_16924\1946049958.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  dict(df1.corr()['latest_price'].abs().sort_values(ascending = False))


{'latest_price': 1.0,
 'taxAssessedValue': 0.8628559938898727,
 'taxAnnualAmount': 0.8156128587000064,
 'bathrooms': 0.551445307264037,
 'resoFactsStats_bathrooms': 0.47586093567052096,
 'livingArea': 0.3919678078297034,
 'resoFactsStats_bathroomsFull': 0.3854093485257164,
 'bedrooms': 0.3236994120035267,
 'resoFactsStats_bedrooms': 0.29769330047900794,
 'schools_2_rating': 0.2879634712029946,
 'schools_1_studentsPerTeacher': 0.2846390962391095,
 'schools_1_rating': 0.24097254565070483,
 'schools_0_rating': 0.23987455502465402,
 'lotSize_value': 0.23968531085209016,
 'schools_1_size': 0.23598605610932682,
 'resoFactsStats_bathroomsHalf': 0.22839099263704998,
 'stories': 0.21967042532231817,
 'numPhotos': 0.181905441001118,
 'longitude': 0.18157016525279043,
 'hasSpa': 0.17628657766491157,
 'coveredSpaces': 0.17196269771541464,
 'parking': 0.16639348500886164,
 'schools_2_studentsPerTeacher': 0.16305886359458596,
 'garageSpaces': 0.16297483664756376,
 'schools_0_distance': 0.15324071928

In [26]:
#Create a "corr_r" row to filter out less correlated features (columns)
df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)

C:\Users\famma\AppData\Local\Temp\ipykernel_16924\4094108136.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)
C:\Users\famma\AppData\Local\Temp\ipykernel_16924\4094108136.py:2: FutureWarning: Behavior when concatenating bool-dtype and numeric-dtype arrays is deprecated; in a future version these will cast to object dtype (instead of coercing bools to numeric values). To retain the old behavior, explicitly cast bool-dtype arrays to numeric dtype.
  df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)


In [27]:
# Check the new row
df1.tail()

Unnamed: 0.1    Unnamed: 0 address_city address_neighborhood  \
16478    16480.00000  22070.000000       Austin                  NaN   
16479    16481.00000  22072.000000       Austin                  NaN   
16480    16482.00000  22073.000000       Austin                  NaN   
16481    16483.00000  22074.000000       Austin                  NaN   
corr_r       0.12465      0.124756          NaN                  NaN   

       address_state address_streetAddress address_subdivision  \
16478             TX     6318 Clairmont Dr                 NaN   
16479             TX      6104 Abilene Trl                 NaN   
16480             TX      7702 Kincheon Ct                 NaN   
16481             TX         5300 Indio Cv                 NaN   
corr_r           NaN                   NaN                 NaN   

        address_zipcode  bathrooms  bedrooms  ...  numOfAccessibilityFeatures  \
16478      78749.000000   2.000000  4.000000  ...                    0.000000   
16479      78749.000000   2.000000  3.000000  ...                    0.000000   
16480      78749.000000   2.500000  3.000000  ...                    0.000000   
16481      78745.000000        NaN       NaN  ...                    0.000000   
corr_r         0.152605   0.551445  0.323699  ...                    0.015614   

       numOfAppliances  numPhotos  numPriceChanges  \
16478         3.000000  33.000000          1.00000   
16479         5.000000  25.000000          5.00000   
16480         1.000000  27.000000          1.00000   
16481         0.000000  11.000000          1.00000   
corr_r        0.048005   0.181905          0.09137   

                                                    photo  \
16478   https://photos.zillowstatic.com/fp/2f26fb52681...   
16479   https://photos.zillowstatic.com/fp/71a9662d303...   
16480   https://photos.zillowstatic.com/fp/f452bc0f4ba...   
16481   https://photos.zillowstatic.com/fp/eb612b3f7c8...   
corr_r                                                NaN   

                                                imagepath  \
16478   /content/gdrive/MyDrive/zillow-images/29487158...   
16479   /content/gdrive/MyDrive/zillow-images/29491564...   
16480   /content/gdrive/MyDrive/zillow-images/29486556...   
16481   /content/gdrive/MyDrive/zillow-images/58314089...   
corr_r                                                NaN   

                                                cleanpath lotSize_value  \
16478   29487158_2f26fb52681290dd3e096cb85fc299ec-p_f.jpg   9321.000000   
16479   29491564_71a9662d3031a1834e29f578f5a92b71-p_f.jpg  10280.000000   
16480   29486556_f452bc0f4ba54fe11c89905773f38840-p_f.jpg   8450.000000   
16481   58314089_eb612b3f7c80976715d8a54e389b8e15-p_f.jpg   7797.000000   
corr_r                                                NaN      0.239685   

       lotSize_unit  lotSize_sqft  
16478          sqft   9321.000000  
16479          sqft  10280.000000  
16480          sqft   8450.000000  
16481          sqft   7797.000000  
corr_r          NaN      0.019782  

[5 rows x 739 columns]

In [28]:
 df1.shape

(16483, 739)

In [29]:
# Exclude columns that have correlation values less than 0.005
df2 = df1.loc[:,df1.loc['corr_r'] >= 0.005]

In [30]:
df2.shape

(16483, 63)

The number of columns are reduced from 736 to 61 by dropping columns that have less than 0.005 correlation value 

In [31]:
df2.head()

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms  dateposted  \
0           0.0        15.0          78732.0        5.0       4.0         NaN   
1           1.0        16.0          78732.0        5.0       5.0         NaN   
2           2.0        17.0          78732.0        3.5       NaN         NaN   
3           3.0        18.0          78732.0        4.0       4.0         NaN   
4           4.0        19.0          78732.0        4.0       4.0         NaN   

    latitude  livingArea  longitude  propertyTaxRate  ...        zpid  \
0  30.354322      4060.0 -97.911278             1.98  ...  70352485.0   
1  30.355553      4558.0 -97.912544             1.98  ...  70352465.0   
2  30.382227      3402.0 -97.908333             1.98  ...  83823478.0   
3  30.352081      4749.0 -97.912048             1.98  ...  70352478.0   
4  30.356226      4867.0 -97.911697             1.98  ...  70352461.0   

   latest_price  latest_salemonth  latest_saleyear  \
0      715000.0               2.0           2018.0   
1     1025000.0               8.0           2020.0   
2       59000.0               3.0           2018.0   
3      825000.0               6.0           2018.0   
4      849000.0               8.0           2019.0   

   numOfAccessibilityFeatures  numOfAppliances  numPhotos  numPriceChanges  \
0                         0.0              3.0       39.0              1.0   
1                         0.0              3.0       58.0              2.0   
2                         0.0              0.0        1.0              1.0   
3                         0.0              1.0       40.0              4.0   
4                         0.0              4.0       40.0              7.0   

   lotSize_value  lotSize_sqft  
0           0.36       15681.6  
1           0.66       28749.6  
2           0.50       21780.0  
3           0.49       21344.4  
4           0.48       20908.8  

[5 rows x 63 columns]

In [32]:
df2.isnull().sum().sort_values(ascending=False)

resoFactsStats_storiesTotal                16446
resoFactsStats_carportSpaces               16427
resoFactsStats_numberOfUnitsInCommunity    16376
resoFactsStats_fireplaces                  16302
resoFactsStats_onMarketDate                16259
                                           ...  
furnished                                      0
longitude                                      0
latitude                                       0
address_zipcode                                0
parking                                        0
Length: 63, dtype: int64

In [33]:
df2_missing_counts = df2.isnull().sum()

In [34]:
df2_columns_to_drop = df2_missing_counts[df2_missing_counts > 14000].index

In [35]:
df2_columns_to_drop

Index(['dateposted', 'resoFactsStats_carportSpaces',
       'resoFactsStats_entryLevel', 'resoFactsStats_fireplaces',
       'resoFactsStats_numberOfUnitsInCommunity',
       'resoFactsStats_onMarketDate', 'resoFactsStats_storiesTotal',
       'resoFactsStats_yearBuiltEffective'],
      dtype='object')

In [36]:
# Create a row called 'null_sum_col' to drop columns that have more than 80% missing values
df2.loc['null_sum_col'] = df2.isnull().sum()

C:\Users\famma\AppData\Local\Temp\ipykernel_16924\3268720471.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2.loc['null_sum_col'] = df2.isnull().sum()


In [37]:
# Filter out columns that has more than 80% missing values
df3 = df2.loc[:, df2.loc['null_sum_col'] < 14000]

In [38]:
df3 = df2.drop(columns=df2_columns_to_drop)

In [39]:
df3.shape

(16484, 55)

In [40]:
df3.head()

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms   latitude  \
0           0.0        15.0          78732.0        5.0       4.0  30.354322   
1           1.0        16.0          78732.0        5.0       5.0  30.355553   
2           2.0        17.0          78732.0        3.5       NaN  30.382227   
3           3.0        18.0          78732.0        4.0       4.0  30.352081   
4           4.0        19.0          78732.0        4.0       4.0  30.356226   

   livingArea  longitude  propertyTaxRate  \
0      4060.0 -97.911278             1.98   
1      4558.0 -97.912544             1.98   
2      3402.0 -97.908333             1.98   
3      4749.0 -97.912048             1.98   
4      4867.0 -97.911697             1.98   

   resoFactsStats_atAGlanceFacts_1_factValue  ...        zpid  latest_price  \
0                                     2007.0  ...  70352485.0      715000.0   
1                                     2007.0  ...  70352465.0     1025000.0   
2                                     2007.0  ...  83823478.0       59000.0   
3                                     2011.0  ...  70352478.0      825000.0   
4                                     2009.0  ...  70352461.0      849000.0   

   latest_salemonth  latest_saleyear  numOfAccessibilityFeatures  \
0               2.0           2018.0                         0.0   
1               8.0           2020.0                         0.0   
2               3.0           2018.0                         0.0   
3               6.0           2018.0                         0.0   
4               8.0           2019.0                         0.0   

   numOfAppliances numPhotos  numPriceChanges lotSize_value lotSize_sqft  
0              3.0      39.0              1.0          0.36      15681.6  
1              3.0      58.0              2.0          0.66      28749.6  
2              0.0       1.0              1.0          0.50      21780.0  
3              1.0      40.0              4.0          0.49      21344.4  
4              4.0      40.0              7.0          0.48      20908.8  

[5 rows x 55 columns]

The number of columns further decreased to 37

In [41]:
# Check how many rows have a total of more than 5 null values
df3[df3.isnull().sum(axis = 1) > 3]

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms  \
2               2.0        17.0          78732.0        3.5       NaN   
11             11.0        26.0          78732.0        NaN       NaN   
78             78.0       106.0          78732.0        NaN       NaN   
93             93.0       122.0          78732.0        NaN       NaN   
104           104.0       136.0          78734.0        3.0       4.0   
...             ...         ...              ...        ...       ...   
16417       16419.0     21857.0          78737.0        3.0       5.0   
16437       16439.0     21946.0          78717.0        3.0       3.0   
16439       16441.0     21951.0          78758.0        1.0       1.0   
16466       16468.0     22039.0          78751.0        NaN       NaN   
16481       16483.0     22074.0          78745.0        NaN       NaN   

        latitude  livingArea  longitude  propertyTaxRate  \
2      30.382227      3402.0 -97.908333             1.98   
11     30.330782      1480.0 -97.925789             1.98   
78     30.382215      2512.0 -97.909950             1.98   
93     30.332327         1.0 -97.911598             1.98   
104    30.342268      3023.0 -97.930733             1.98   
...          ...         ...        ...              ...   
16417  30.198322      2747.0 -97.989227             2.01   
16437  30.476957      1645.0 -97.772057             2.21   
16439  30.406792       657.0 -97.697861             1.98   
16466  30.311029      3700.0 -97.719620             1.98   
16481  30.222555      3120.0 -97.805099             1.98   

       resoFactsStats_atAGlanceFacts_1_factValue  ...          zpid  \
2                                         2007.0  ...  8.382348e+07   
11                                        1975.0  ...  2.934395e+07   
78                                        2006.0  ...  7.998023e+07   
93                                        2005.0  ...  2.419302e+08   
104                                       1992.0  ...  5.829815e+07   
...                                          ...  ...           ...   
16417                                     2006.0  ...  8.007668e+07   
16437                                     2018.0  ...  2.093629e+09   
16439                                     1986.0  ...  2.113423e+09   
16466                                     2020.0  ...  2.940159e+07   
16481                                     1980.0  ...  5.831409e+07   

       latest_price  latest_salemonth  latest_saleyear  \
2           59000.0               3.0           2018.0   
11         315000.0               4.0           2019.0   
78         100000.0               2.0           2018.0   
93        2195000.0               4.0           2019.0   
104       1500000.0               8.0           2020.0   
...             ...               ...              ...   
16417      365000.0               1.0           2018.0   
16437      331700.0               1.0           2018.0   
16439      124999.0               9.0           2019.0   
16466     1299900.0              12.0           2020.0   
16481      515000.0               1.0           2018.0   

       numOfAccessibilityFeatures  numOfAppliances numPhotos  numPriceChanges  \
2                             0.0              0.0       1.0              1.0   
11                            0.0              1.0      18.0              9.0   
78                            0.0              0.0      15.0              1.0   
93                            0.0              0.0       5.0              1.0   
104                           0.0              0.0      31.0              2.0   
...                           ...              ...       ...              ...   
16417                         0.0              7.0      14.0              1.0   
16437                         0.0              0.0      16.0              3.0   
16439                         0.0              5.0      17.0              7.0   
16466                         0.0              0.

In [42]:
df3['null_sum_rows'] = df3.isnull().sum(axis = 1)

In [43]:
df3.head()

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms   latitude  \
0           0.0        15.0          78732.0        5.0       4.0  30.354322   
1           1.0        16.0          78732.0        5.0       5.0  30.355553   
2           2.0        17.0          78732.0        3.5       NaN  30.382227   
3           3.0        18.0          78732.0        4.0       4.0  30.352081   
4           4.0        19.0          78732.0        4.0       4.0  30.356226   

   livingArea  longitude  propertyTaxRate  \
0      4060.0 -97.911278             1.98   
1      4558.0 -97.912544             1.98   
2      3402.0 -97.908333             1.98   
3      4749.0 -97.912048             1.98   
4      4867.0 -97.911697             1.98   

   resoFactsStats_atAGlanceFacts_1_factValue  ...  latest_price  \
0                                     2007.0  ...      715000.0   
1                                     2007.0  ...     1025000.0   
2                                     2007.0  ...       59000.0   
3                                     2011.0  ...      825000.0   
4                                     2009.0  ...      849000.0   

   latest_salemonth  latest_saleyear  numOfAccessibilityFeatures  \
0               2.0           2018.0                         0.0   
1               8.0           2020.0                         0.0   
2               3.0           2018.0                         0.0   
3               6.0           2018.0                         0.0   
4               8.0           2019.0                         0.0   

   numOfAppliances  numPhotos numPriceChanges  lotSize_value lotSize_sqft  \
0              3.0       39.0             1.0           0.36      15681.6   
1              3.0       58.0             2.0           0.66      28749.6   
2              0.0        1.0             1.0           0.50      21780.0   
3              1.0       40.0             4.0           0.49      21344.4   
4              4.0       40.0             7.0           0.48      20908.8   

  null_sum_rows  
0             1  
1             0  
2             7  
3             1  
4             1  

[5 rows x 56 columns]

In [44]:
df3.shape

(16484, 56)

In [45]:
df3 = df3[df3['null_sum_rows'] <= 3]

In [46]:
df3.shape

(14735, 56)

In [47]:
# Drop unwanated columns 
df3 = df3.drop(columns=['Unnamed: 0.1', 'Unnamed: 0', 'numPhotos', 'null_sum_rows','lotSize_value'])

In [48]:
# Drop rows that are no longer needed
df3.drop(index=['corr_r', 'null_sum_col'], inplace = True)

In [49]:
df3.tail()

address_zipcode  bathrooms  bedrooms   latitude  livingArea  longitude  \
16476          78749.0        2.0       3.0  30.224886      1496.0 -97.862335   
16477          78749.0        2.0       3.0  30.221336      2270.0 -97.879677   
16478          78749.0        2.0       4.0  30.216406      2645.0 -97.876045   
16479          78749.0        2.0       3.0  30.213816      1469.0 -97.873711   
16480          78749.0        2.5       3.0  30.218878      1700.0 -97.864960   

       propertyTaxRate  resoFactsStats_atAGlanceFacts_1_factValue  \
16476             1.98                                     1988.0   
16477             1.98                                     1999.0   
16478             1.98                                     1993.0   
16479             1.98                                     1992.0   
16480             1.98                                     1986.0   

       resoFactsStats_bathrooms  resoFactsStats_bathroomsFull  ...  \
16476                       2.0                           0.0  ...   
16477                       2.0                           0.0  ...   
16478                       2.0                           0.0  ...   
16479                       2.0                           2.0  ...   
16480                       NaN                           0.0  ...   

       schools_2_totalCount  yearBuilt        zpid  latest_price  \
16476                   1.0     1988.0  29483690.0      292500.0   
16477                   1.0     1999.0  29484499.0      425000.0   
16478                   1.0     1993.0  29487158.0      424999.0   
16479                   1.0     1992.0  29491564.0      316000.0   
16480                   1.0     1986.0  29486556.0      384900.0   

      latest_salemonth  latest_saleyear numOfAccessibilityFeatures  \
16476              1.0           2018.0                        0.0   
16477              1.0           2018.0                        0.0   
16478              1.0           2018.0                        0.0   
16479              1.0           2018.0                        0.0   
16480              1.0           2018.0                        0.0   

      numOfAppliances numPriceChanges lotSize_sqft  
16476             0.0             1.0       5662.0  
16477             3.0             1.0      23522.4  
16478             3.0             1.0       9321.0  
16479             5.0             5.0      10280.0  
16480             1.0             1.0       8450.0  

[5 rows x 51 columns]

In [50]:
df3['price_per_sqft'] = df3['latest_price'] / df3['livingArea']

In [51]:
df3.shape

(14733, 52)

In [52]:
# Evaluate the amount of missing values in each columns
df3.isnull().sum()

address_zipcode                                 0
bathrooms                                     118
bedrooms                                       45
latitude                                        0
livingArea                                      0
longitude                                       0
propertyTaxRate                                 0
resoFactsStats_atAGlanceFacts_1_factValue       0
resoFactsStats_bathrooms                      158
resoFactsStats_bathroomsFull                    0
resoFactsStats_bathroomsHalf                    0
resoFactsStats_bathroomsThreeQuarter          189
resoFactsStats_bedrooms                         4
coveredSpaces                                8315
furnished                                       0
garageSpaces                                    0
hasAttachedGarage                               0
hasCarport                                      0
hasCooling                                      0
hasGarage                                       0


In [53]:
# Evaluate the amount of missing values across rows
df3[df3.isnull().sum(axis = 1) > 0]

address_zipcode  bathrooms  bedrooms   latitude  livingArea  longitude  \
0              78732.0        5.0       4.0  30.354322      4060.0 -97.911278   
3              78732.0        4.0       4.0  30.352081      4749.0 -97.912048   
4              78732.0        4.0       4.0  30.356226      4867.0 -97.911697   
6              78732.0        NaN       3.0  30.329847      2326.0 -97.926910   
7              78732.0        5.0       4.0  30.352270      3385.0 -97.912140   
...                ...        ...       ...        ...         ...        ...   
16476          78749.0        2.0       3.0  30.224886      1496.0 -97.862335   
16477          78749.0        2.0       3.0  30.221336      2270.0 -97.879677   
16478          78749.0        2.0       4.0  30.216406      2645.0 -97.876045   
16479          78749.0        2.0       3.0  30.213816      1469.0 -97.873711   
16480          78749.0        2.5       3.0  30.218878      1700.0 -97.864960   

       propertyTaxRate  resoFactsStats_atAGlanceFacts_1_factValue  \
0                 1.98                                     2007.0   
3                 1.98                                     2011.0   
4                 1.98                                     2009.0   
6                 1.98                                     2018.0   
7                 1.98                                     2007.0   
...                ...                                        ...   
16476             1.98                                     1988.0   
16477             1.98                                     1999.0   
16478             1.98                                     1993.0   
16479             1.98                                     1992.0   
16480             1.98                                     1986.0   

       resoFactsStats_bathrooms  resoFactsStats_bathroomsFull  ...  yearBuilt  \
0                           5.0                           4.0  ...     2007.0   
3                           4.0                           3.0  ...     2011.0   
4                           4.0                           4.0  ...     2009.0   
6                           0.0                           0.0  ...     2018.0   
7                           5.0                           4.0  ...     2007.0   
...                         ...                           ...  ...        ...   
16476                       2.0                           0.0  ...     1988.0   
16477                       2.0                           0.0  ...     1999.0   
16478                       2.0                           0.0  ...     1993.0   
16479                       2.0                           2.0  ...     1992.0   
16480                       NaN                           0.0  ...     1986.0   

              zpid  latest_price  latest_salemonth latest_saleyear  \
0       70352485.0      715000.0               2.0          2018.0   
3       70352478.0      825000.0               6.0          2018.0   
4       70352461.0      849000.0               8.0          2019.0   
6      124904977.0      139990.0               4.0          2018.0   
7       70352477.0      800000.0              10.0          2020.0   
...            ...           ...               ...             ...   
16476   29483690.0      292500.0               1.0          2018.0   
16477   29484499.0      425000.0               1.0          2018.0   
16478   29487158.0      424999.0               1.0          2018.0   
16479   29491564.0      316000.0               1.0          2018.0   
16480   29486556.0      384900.0               1.0          2018.0   

       numOfAccessibilityFeatures numOfAppliances numPriceChanges  \
0                             0.0             3.0             1.0   
3                             0.0             1.0             4.0   
4                             0.0             4.0             7.0   
6                             0.0             0.0             1.0   
7                             0.0             

In [54]:
df3.mean().round(1)

address_zipcode                                 78735.9
bathrooms                                           2.7
bedrooms                                            3.4
latitude                                           30.3
livingArea                                       2153.2
longitude                                         -97.8
propertyTaxRate                                     2.0
resoFactsStats_atAGlanceFacts_1_factValue        1987.5
resoFactsStats_bathrooms                            2.6
resoFactsStats_bathroomsFull                        2.0
resoFactsStats_bathroomsHalf                        0.4
resoFactsStats_bathroomsThreeQuarter                0.0
resoFactsStats_bedrooms                             3.4
coveredSpaces                                       1.7
furnished                                           0.0
garageSpaces                                        1.2
hasAttachedGarage                                   0.0
hasCarport                                      

In [55]:
df3.shape

(14733, 52)

In [56]:
(df3.address_zipcode.min(), df3.address_zipcode.max())

(78617.0, 78759.0)

In [57]:
df3.address_zipcode.dtype

dtype('float64')

In [58]:
df3['zipcode'] = df3.address_zipcode.astype('int')

In [59]:
df3 = df3.set_index(['zipcode'])

In [60]:
df3.head()

address_zipcode  bathrooms  bedrooms   latitude  livingArea  \
zipcode                                                                
78732            78732.0        5.0       4.0  30.354322      4060.0   
78732            78732.0        5.0       5.0  30.355553      4558.0   
78732            78732.0        4.0       4.0  30.352081      4749.0   
78732            78732.0        4.0       4.0  30.356226      4867.0   
78732            78732.0        4.0       5.0  30.341896      3485.0   

         longitude  propertyTaxRate  \
zipcode                               
78732   -97.911278             1.98   
78732   -97.912544             1.98   
78732   -97.912048             1.98   
78732   -97.911697             1.98   
78732   -97.907944             1.98   

         resoFactsStats_atAGlanceFacts_1_factValue  resoFactsStats_bathrooms  \
zipcode                                                                        
78732                                       2007.0                       5.0   
78732                                       2007.0                       5.0   
78732                                       2011.0                       4.0   
78732                                       2009.0                       4.0   
78732                                       2009.0                       4.0   

         resoFactsStats_bathroomsFull  ...  yearBuilt        zpid  \
zipcode                                ...                          
78732                             4.0  ...     2007.0  70352485.0   
78732                             5.0  ...     2007.0  70352465.0   
78732                             3.0  ...     2011.0  70352478.0   
78732                             4.0  ...     2009.0  70352461.0   
78732                             3.0  ...     2009.0  89028960.0   

         latest_price  latest_salemonth latest_saleyear  \
zipcode                                                   
78732        715000.0               2.0          2018.0   
78732       1025000.0               8.0          2020.0   
78732        825000.0               6.0          2018.0   
78732        849000.0               8.0          2019.0   
78732        625000.0               7.0          2019.0   

         numOfAccessibilityFeatures numOfAppliances numPriceChanges  \
zipcode                                                               
78732                           0.0             3.0             1.0   
78732                           0.0             3.0             2.0   
78732                           0.0             1.0             4.0   
78732                           0.0             4.0             7.0   
78732                           0.0             4.0             4.0   

        lotSize_sqft price_per_sqft  
zipcode                              
78732        15681.6     176.108374  
78732        28749.6     224.879333  
78732        21344.4     173.720783  
78732        20908.8     174.440107  
78732         7797.0     179.340029  

[5 rows x 52 columns]

# load the US Income data for Austin and merge it with Austin House Listings cleaned data

In [61]:
# load the us_income data
df1_inc = pd.read_csv('../01_raw_data/us_income_zipcode.csv', low_memory=False)

In [62]:
df1_inc.shape

(364998, 111)

In [63]:
df2_inc = df1_inc[df1_inc['ZIP'].between(78617,78759)]

In [64]:
df2_inc.shape

(1025, 111)

In [65]:
list(df2_inc.columns)

['ZIP',
 'Geography',
 'Geographic Area Name',
 'Households',
 'Households Margin of Error',
 'Households Less Than $10,000',
 'Households Less Than $10,000 Margin of Error',
 'Households $10,000 to $14,999',
 'Households $10,000 to $14,999 Margin of Error',
 'Households $15,000 to $24,999',
 'Households $15,000 to $24,999 Margin of Error',
 'Households $25,000 to $34,999',
 'Households $25,000 to $34,999 Margin of Error',
 'Households $35,000 to $49,999',
 'Households $35,000 to $49,999 Margin of Error',
 'Households $50,000 to $74,999',
 'Households $50,000 to $74,999 Margin of Error',
 'Households $75,000 to $99,999',
 'Households $75,000 to $99,999 Margin of Error',
 'Households $100,000 to $149,999',
 'Households $100,000 to $149,999 Margin of Error',
 'Households $150,000 to $199,999',
 'Households $150,000 to $199,999 Margin of Error',
 'Households $200,000 or More',
 'Households $200,000 or More Margin of Error',
 'Households Median Income (Dollars)',
 'Households Median Income

In [66]:
df2_inc = df2_inc.set_index(['ZIP'])

In [67]:
df2_inc.head()

Geography Geographic Area Name  Households  \
ZIP                                                      
78617  860Z200US78617          ZCTA5 78617      7662.0   
78618  860Z200US78618          ZCTA5 78618       129.0   
78619  860Z200US78619          ZCTA5 78619      2159.0   
78620  860Z200US78620          ZCTA5 78620      6538.0   
78621  860Z200US78621          ZCTA5 78621      8103.0   

       Households Margin of Error  Households Less Than $10,000  \
ZIP                                                               
78617                       682.0                           3.0   
78618                        50.0                           6.2   
78619                       372.0                           0.0   
78620                       432.0                           1.5   
78621                       552.0                           5.0   

       Households Less Than $10,000 Margin of Error  \
ZIP                                                   
78617                                           2.3   
78618                                           6.8   
78619                                           2.2   
78620                                           1.0   
78621                                           2.3   

       Households $10,000 to $14,999  \
ZIP                                    
78617                            1.7   
78618                            0.8   
78619                            1.0   
78620                            3.4   
78621                            2.2   

       Households $10,000 to $14,999 Margin of Error  \
ZIP                                                    
78617                                            2.0   
78618                                            2.4   
78619                                            1.2   
78620                                            2.1   
78621                                            1.4   

       Households $15,000 to $24,999  \
ZIP                                    
78617                            3.7   
78618                            5.4   
78619                            0.0   
78620                            2.9   
78621                            2.4   

       Households $15,000 to $24,999 Margin of Error  ...  \
ZIP                                                   ...   
78617                                            1.8  ...   
78618                                            5.5  ...   
78619                                            2.2  ...   
78620                                            1.4  ...   
78621                                            1.4  ...   

       Nonfamily Households $150,000 to $199,999  \
ZIP                                                
78617                                        2.1   
78618                                       14.8   
78619                                        0.0   
78620                                        4.8   
78621                                       12.2   

       Nonfamily Households $150,000 to $199,999 Margin of Error  \
ZIP                                                                
78617                                                3.4           
78618                                               22.7           
78619                                               25.9           
78620                                                3.9           
78621                                               15.1           

       Nonfamily Households $200,000 or More  \
ZIP                                            
78617                                    2.6   
78618                                   14.8   
78619                                    0.0   
78620                                    5.4   
78621                                    2.0   

       Nonfamily Households $200,000 or More Margin of Error  \
ZIP                                                            
78617                                                4.0   

In [68]:
#Analyze Data across a row
dict(df2_inc.iloc[0])

{'Geography': '860Z200US78617',
 'Geographic Area Name': 'ZCTA5 78617',
 'Households': 7662.0,
 'Households Margin of Error': 682.0,
 'Households Less Than $10,000': 3.0,
 'Households Less Than $10,000 Margin of Error': 2.3,
 'Households $10,000 to $14,999': 1.7,
 'Households $10,000 to $14,999 Margin of Error': 2.0,
 'Households $15,000 to $24,999': 3.7,
 'Households $15,000 to $24,999 Margin of Error': 1.8,
 'Households $25,000 to $34,999': 6.6,
 'Households $25,000 to $34,999 Margin of Error': 2.4,
 'Households $35,000 to $49,999': 19.2,
 'Households $35,000 to $49,999 Margin of Error': 6.4,
 'Households $50,000 to $74,999': 20.1,
 'Households $50,000 to $74,999 Margin of Error': 4.9,
 'Households $75,000 to $99,999': 10.3,
 'Households $75,000 to $99,999 Margin of Error': 3.9,
 'Households $100,000 to $149,999': 23.0,
 'Households $100,000 to $149,999 Margin of Error': 5.4,
 'Households $150,000 to $199,999': 6.0,
 'Households $150,000 to $199,999 Margin of Error': 2.8,
 'Household

In [69]:
selected_cols = [col for col in df2_inc.columns if 'Mean Income' in col and 'Margin of Error' not in col]

In [70]:
df2_inc = df2_inc[selected_cols]

In [71]:
df2_inc.head()

Households Mean Income (Dollars)  Families Mean Income (Dollars)  \
ZIP                                                                       
78617                           99354.0                        106461.0   
78618                          105468.0                         92582.0   
78619                          165887.0                        169968.0   
78620                          151574.0                        165526.0   
78621                           94097.0                         95930.0   

       Married-Couple Families Mean Income (Dollars)  \
ZIP                                                    
78617                                            NaN   
78618                                            NaN   
78619                                            NaN   
78620                                            NaN   
78621                                            NaN   

       Nonfamily Households Mean Income (Dollars)  
ZIP                                                
78617                                     56019.0  
78618                                    154152.0  
78619                                     75649.0  
78620                                     70972.0  
78621                                     80298.0

In [72]:
df2_inc.shape

(1025, 4)

In [73]:
df2_inc_average = df2_inc.mean(axis = 1).round(1)

In [74]:
df2_inc['zip_inc_ave'] = df2_inc.mean(axis = 1).round(2)

In [75]:
df2_inc.columns

Index(['Households Mean Income (Dollars)', 'Families Mean Income (Dollars)',
       'Married-Couple Families Mean Income (Dollars)',
       'Nonfamily Households Mean Income (Dollars)', 'zip_inc_ave'],
      dtype='object')

In [76]:
df2_inc.rename(columns={'ZIP': 'zipcode', '': 'inc_average'}, inplace=True)
df2_inc.head()

Households Mean Income (Dollars)  Families Mean Income (Dollars)  \
ZIP                                                                       
78617                           99354.0                        106461.0   
78618                          105468.0                         92582.0   
78619                          165887.0                        169968.0   
78620                          151574.0                        165526.0   
78621                           94097.0                         95930.0   

       Married-Couple Families Mean Income (Dollars)  \
ZIP                                                    
78617                                            NaN   
78618                                            NaN   
78619                                            NaN   
78620                                            NaN   
78621                                            NaN   

       Nonfamily Households Mean Income (Dollars)  zip_inc_ave  
ZIP                                                             
78617                                     56019.0     87278.00  
78618                                    154152.0    117400.67  
78619                                     75649.0    137168.00  
78620                                     70972.0    129357.33  
78621                                     80298.0     90108.33

In [77]:
df3.head()

address_zipcode  bathrooms  bedrooms   latitude  livingArea  \
zipcode                                                                
78732            78732.0        5.0       4.0  30.354322      4060.0   
78732            78732.0        5.0       5.0  30.355553      4558.0   
78732            78732.0        4.0       4.0  30.352081      4749.0   
78732            78732.0        4.0       4.0  30.356226      4867.0   
78732            78732.0        4.0       5.0  30.341896      3485.0   

         longitude  propertyTaxRate  \
zipcode                               
78732   -97.911278             1.98   
78732   -97.912544             1.98   
78732   -97.912048             1.98   
78732   -97.911697             1.98   
78732   -97.907944             1.98   

         resoFactsStats_atAGlanceFacts_1_factValue  resoFactsStats_bathrooms  \
zipcode                                                                        
78732                                       2007.0                       5.0   
78732                                       2007.0                       5.0   
78732                                       2011.0                       4.0   
78732                                       2009.0                       4.0   
78732                                       2009.0                       4.0   

         resoFactsStats_bathroomsFull  ...  yearBuilt        zpid  \
zipcode                                ...                          
78732                             4.0  ...     2007.0  70352485.0   
78732                             5.0  ...     2007.0  70352465.0   
78732                             3.0  ...     2011.0  70352478.0   
78732                             4.0  ...     2009.0  70352461.0   
78732                             3.0  ...     2009.0  89028960.0   

         latest_price  latest_salemonth latest_saleyear  \
zipcode                                                   
78732        715000.0               2.0          2018.0   
78732       1025000.0               8.0          2020.0   
78732        825000.0               6.0          2018.0   
78732        849000.0               8.0          2019.0   
78732        625000.0               7.0          2019.0   

         numOfAccessibilityFeatures numOfAppliances numPriceChanges  \
zipcode                                                               
78732                           0.0             3.0             1.0   
78732                           0.0             3.0             2.0   
78732                           0.0             1.0             4.0   
78732                           0.0             4.0             7.0   
78732                           0.0             4.0             4.0   

        lotSize_sqft price_per_sqft  
zipcode                              
78732        15681.6     176.108374  
78732        28749.6     224.879333  
78732        21344.4     173.720783  
78732        20908.8     174.440107  
78732         7797.0     179.340029  

[5 rows x 52 columns]

In [78]:
df3.shape

(14733, 52)

In [79]:
df2_inc_average.head()

ZIP
78617     87278.0
78618    117400.7
78619    137168.0
78620    129357.3
78621     90108.3
dtype: float64

In [80]:
df2_inc_average.tail()

ZIP
78754    60590.3
78756    84134.3
78757    67990.0
78758    52497.7
78759    91688.0
dtype: float64

In [81]:
df2_inc_average.shape

(1025,)

In [82]:
type(df2_inc_average)

pandas.core.series.Series

In [83]:
df3_inc_average = pd.DataFrame(df2_inc_average)
type(df3_inc_average)

pandas.core.frame.DataFrame

In [84]:
df3_inc_average.head(2)

0
ZIP            
78617   87278.0
78618  117400.7

In [85]:
df4 = df3.reset_index(level=None, drop=False, inplace=False, col_level=0, col_fill="")
df4.head()

zipcode  address_zipcode  bathrooms  bedrooms   latitude  livingArea  \
0    78732          78732.0        5.0       4.0  30.354322      4060.0   
1    78732          78732.0        5.0       5.0  30.355553      4558.0   
2    78732          78732.0        4.0       4.0  30.352081      4749.0   
3    78732          78732.0        4.0       4.0  30.356226      4867.0   
4    78732          78732.0        4.0       5.0  30.341896      3485.0   

   longitude  propertyTaxRate  resoFactsStats_atAGlanceFacts_1_factValue  \
0 -97.911278             1.98                                     2007.0   
1 -97.912544             1.98                                     2007.0   
2 -97.912048             1.98                                     2011.0   
3 -97.911697             1.98                                     2009.0   
4 -97.907944             1.98                                     2009.0   

   resoFactsStats_bathrooms  ...  yearBuilt        zpid  latest_price  \
0                       5.0  ...     2007.0  70352485.0      715000.0   
1                       5.0  ...     2007.0  70352465.0     1025000.0   
2                       4.0  ...     2011.0  70352478.0      825000.0   
3                       4.0  ...     2009.0  70352461.0      849000.0   
4                       4.0  ...     2009.0  89028960.0      625000.0   

   latest_salemonth  latest_saleyear numOfAccessibilityFeatures  \
0               2.0           2018.0                        0.0   
1               8.0           2020.0                        0.0   
2               6.0           2018.0                        0.0   
3               8.0           2019.0                        0.0   
4               7.0           2019.0                        0.0   

   numOfAppliances numPriceChanges lotSize_sqft price_per_sqft  
0              3.0             1.0      15681.6     176.108374  
1              3.0             2.0      28749.6     224.879333  
2              1.0             4.0      21344.4     173.720783  
3              4.0             7.0      20908.8     174.440107  
4              4.0             4.0       7797.0     179.340029  

[5 rows x 53 columns]

In [86]:
df4_inc_average = df3_inc_average.reset_index(level=None, drop=False, inplace=False, col_level=0, col_fill="")
df4_inc_average.head() 

ZIP         0
0  78617   87278.0
1  78618  117400.7
2  78619  137168.0
3  78620  129357.3
4  78621   90108.3

In [87]:
df4_inc_average.rename(columns={'ZIP': 'zipcode', '0': 'average_income'}, inplace=True)
df4_inc_average.head()

zipcode         0
0    78617   87278.0
1    78618  117400.7
2    78619  137168.0
3    78620  129357.3
4    78621   90108.3

In [88]:
df4_inc_average.shape

(1025, 2)

In [89]:
df4.shape

(14733, 53)

In [90]:
merged_df = pd.merge(df4, df4_inc_average, on='zipcode', how='left')
merged_df.shape

C:\Users\famma\AppData\Local\Temp\ipykernel_16924\4228725122.py:1: FutureWarning: In a future version, object-dtype columns with all-bool values will not be included in reductions with bool_only=True. Explicitly cast to bool dtype instead.
  merged_df = pd.merge(df4, df4_inc_average, on='zipcode', how='left')


(162063, 54)

In [91]:
merged_df.head()

zipcode  address_zipcode  bathrooms  bedrooms   latitude  livingArea  \
0    78732          78732.0        5.0       4.0  30.354322      4060.0   
1    78732          78732.0        5.0       4.0  30.354322      4060.0   
2    78732          78732.0        5.0       4.0  30.354322      4060.0   
3    78732          78732.0        5.0       4.0  30.354322      4060.0   
4    78732          78732.0        5.0       4.0  30.354322      4060.0   

   longitude  propertyTaxRate  resoFactsStats_atAGlanceFacts_1_factValue  \
0 -97.911278             1.98                                     2007.0   
1 -97.911278             1.98                                     2007.0   
2 -97.911278             1.98                                     2007.0   
3 -97.911278             1.98                                     2007.0   
4 -97.911278             1.98                                     2007.0   

   resoFactsStats_bathrooms  ...        zpid  latest_price  latest_salemonth  \
0                       5.0  ...  70352485.0      715000.0               2.0   
1                       5.0  ...  70352485.0      715000.0               2.0   
2                       5.0  ...  70352485.0      715000.0               2.0   
3                       5.0  ...  70352485.0      715000.0               2.0   
4                       5.0  ...  70352485.0      715000.0               2.0   

   latest_saleyear  numOfAccessibilityFeatures numOfAppliances  \
0           2018.0                         0.0             3.0   
1           2018.0                         0.0             3.0   
2           2018.0                         0.0             3.0   
3           2018.0                         0.0             3.0   
4           2018.0                         0.0             3.0   

   numPriceChanges lotSize_sqft price_per_sqft         0  
0              1.0      15681.6     176.108374  194007.3  
1              1.0      15681.6     176.108374  195901.3  
2              1.0      15681.6     176.108374  186103.0  
3              1.0      15681.6     176.108374  172963.7  
4              1.0      15681.6     176.108374  162950.7  

[5 rows x 54 columns]

In [92]:
merged_df.rename(columns={'ZIP': 'zipcode', '0': 'average_income'}, inplace=True)
merged_df.head()

zipcode  address_zipcode  bathrooms  bedrooms   latitude  livingArea  \
0    78732          78732.0        5.0       4.0  30.354322      4060.0   
1    78732          78732.0        5.0       4.0  30.354322      4060.0   
2    78732          78732.0        5.0       4.0  30.354322      4060.0   
3    78732          78732.0        5.0       4.0  30.354322      4060.0   
4    78732          78732.0        5.0       4.0  30.354322      4060.0   

   longitude  propertyTaxRate  resoFactsStats_atAGlanceFacts_1_factValue  \
0 -97.911278             1.98                                     2007.0   
1 -97.911278             1.98                                     2007.0   
2 -97.911278             1.98                                     2007.0   
3 -97.911278             1.98                                     2007.0   
4 -97.911278             1.98                                     2007.0   

   resoFactsStats_bathrooms  ...        zpid  latest_price  latest_salemonth  \
0                       5.0  ...  70352485.0      715000.0               2.0   
1                       5.0  ...  70352485.0      715000.0               2.0   
2                       5.0  ...  70352485.0      715000.0               2.0   
3                       5.0  ...  70352485.0      715000.0               2.0   
4                       5.0  ...  70352485.0      715000.0               2.0   

   latest_saleyear  numOfAccessibilityFeatures numOfAppliances  \
0           2018.0                         0.0             3.0   
1           2018.0                         0.0             3.0   
2           2018.0                         0.0             3.0   
3           2018.0                         0.0             3.0   
4           2018.0                         0.0             3.0   

   numPriceChanges lotSize_sqft price_per_sqft         0  
0              1.0      15681.6     176.108374  194007.3  
1              1.0      15681.6     176.108374  195901.3  
2              1.0      15681.6     176.108374  186103.0  
3              1.0      15681.6     176.108374  172963.7  
4              1.0      15681.6     176.108374  162950.7  

[5 rows x 54 columns]

In [93]:
merged_df.rename(columns={merged_df.columns[-1]: 'income_ave'}, inplace=True)
merged_df.head()

zipcode  address_zipcode  bathrooms  bedrooms   latitude  livingArea  \
0    78732          78732.0        5.0       4.0  30.354322      4060.0   
1    78732          78732.0        5.0       4.0  30.354322      4060.0   
2    78732          78732.0        5.0       4.0  30.354322      4060.0   
3    78732          78732.0        5.0       4.0  30.354322      4060.0   
4    78732          78732.0        5.0       4.0  30.354322      4060.0   

   longitude  propertyTaxRate  resoFactsStats_atAGlanceFacts_1_factValue  \
0 -97.911278             1.98                                     2007.0   
1 -97.911278             1.98                                     2007.0   
2 -97.911278             1.98                                     2007.0   
3 -97.911278             1.98                                     2007.0   
4 -97.911278             1.98                                     2007.0   

   resoFactsStats_bathrooms  ...        zpid  latest_price  latest_salemonth  \
0                       5.0  ...  70352485.0      715000.0               2.0   
1                       5.0  ...  70352485.0      715000.0               2.0   
2                       5.0  ...  70352485.0      715000.0               2.0   
3                       5.0  ...  70352485.0      715000.0               2.0   
4                       5.0  ...  70352485.0      715000.0               2.0   

   latest_saleyear  numOfAccessibilityFeatures numOfAppliances  \
0           2018.0                         0.0             3.0   
1           2018.0                         0.0             3.0   
2           2018.0                         0.0             3.0   
3           2018.0                         0.0             3.0   
4           2018.0                         0.0             3.0   

   numPriceChanges lotSize_sqft price_per_sqft income_ave  
0              1.0      15681.6     176.108374   194007.3  
1              1.0      15681.6     176.108374   195901.3  
2              1.0      15681.6     176.108374   186103.0  
3              1.0      15681.6     176.108374   172963.7  
4              1.0      15681.6     176.108374   162950.7  

[5 rows x 54 columns]

In [94]:
merged_df.columns

Index(['zipcode', 'address_zipcode', 'bathrooms', 'bedrooms', 'latitude',
       'livingArea', 'longitude', 'propertyTaxRate',
       'resoFactsStats_atAGlanceFacts_1_factValue', 'resoFactsStats_bathrooms',
       'resoFactsStats_bathroomsFull', 'resoFactsStats_bathroomsHalf',
       'resoFactsStats_bathroomsThreeQuarter', 'resoFactsStats_bedrooms',
       'coveredSpaces', 'furnished', 'garageSpaces', 'hasAttachedGarage',
       'hasCarport', 'hasCooling', 'hasGarage', 'hasHeating', 'hasSpa',
       'hasView', 'parking', 'stories', 'taxAnnualAmount', 'taxAssessedValue',
       'resoFactsStats_yearBuilt', 'schools_0_distance', 'schools_0_rating',
       'schools_0_size', 'schools_0_studentsPerTeacher',
       'schools_0_totalCount', 'schools_1_distance', 'schools_1_rating',
       'schools_1_size', 'schools_1_studentsPerTeacher',
       'schools_1_totalCount', 'schools_2_distance', 'schools_2_rating',
       'schools_2_studentsPerTeacher', 'schools_2_totalCount', 'yearBuilt',
       'zp

In [103]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 162063 entries, 0 to 162062
Data columns (total 54 columns):
 #   Column                                     Non-Null Count   Dtype  
---  ------                                     --------------   -----  
 0   zipcode                                    162063 non-null  int64  
 1   address_zipcode                            162063 non-null  float64
 2   bathrooms                                  160765 non-null  float64
 3   bedrooms                                   161568 non-null  float64
 4   latitude                                   162063 non-null  float64
 5   livingArea                                 162063 non-null  float64
 6   longitude                                  162063 non-null  float64
 7   propertyTaxRate                            162063 non-null  float64
 8   resoFactsStats_atAGlanceFacts_1_factValue  162063 non-null  float64
 9   resoFactsStats_bathrooms                   160325 non-null  float64
 10  resoFact

In [108]:
# Save the DataFrame to a CSV file
merged_df.to_csv('../03_processed_data/austin_housePrice_and_income_data.csv', index=False)

print("DataFrame saved to 'austin_housePrice_and_income_data.csv'.")

#datapath = '../03_processed_data'
#save_file(merged_df, 'austin_housePrice_and_income_data.csv', datapath)

DataFrame saved to 'austin_housePrice_and_income_data.csv'.
